RecoLocalCalo/HcalRecAlgos/src/DoMahiAlgo.cc

```diff
@@ -0,0 +1,563 @@
+
+    if (offset==100) continue;		       
+    else { 
+      invCovMat_ += ampVec_.coeff(iBX)*ampVec_.coeff(iBX)
+	*pulseCovArray_.at(offset+BXOffset_).block(FullTSOffset_-offset, FullTSOffset_-offset, TSSize_, TSSize_);
+    }
+  }
+  
+  covDecomp_.compute(invCovMat_);
+  
+  return status;
+}
+
+bool DoMahiAlgo::NNLS() {
+  const unsigned int npulse = nPulseTot_;
+  
+  for (unsigned int iBX=0; iBX<npulse; iBX++) {
+    int offset=bxs_.coeff(iBX);
+    if (offset==100) {
+      pulseMat_.col(iBX) = SampleVector::Ones();
+    }
+    else {
+      pulseMat_.col(iBX) = pulseShapeArray_.at(offset+BXOffset_).segment(FullTSOffset_-offset, TSSize_);
+    }
+  }
+
+  invcovp_ = covDecomp_.matrixL().solve(pulseMat_);
+  aTaMat_ = invcovp_.transpose().lazyProduct(invcovp_);
+  aTbVec_ = invcovp_.transpose().lazyProduct(covDecomp_.matrixL().solve(amplitudes_));
+  
+  int iter = 0;
+  Index idxwmax = 0;
+  double wmax = 0.0;
+  double threshold = nnlsThresh_;
+  
+  while (true) {    
+    if (iter>0 || nP_==0) {
+      if ( nP_==std::min(npulse, TSSize_)) break;
+      
+      const unsigned int nActive = npulse - nP_;
+      updateWork_ = aTbVec_ - aTaMat_*ampVec_;
+      
+      Index idxwmaxprev = idxwmax;
+      double wmaxprev = wmax;
+      wmax = updateWork_.tail(nActive).maxCoeff(&idxwmax);
+      
+      if (wmax<threshold || (idxwmax==idxwmaxprev && wmax==wmaxprev)) {
+	break;
+      }
+      
+      if (iter>=nMaxItersNNLS_) {
+	break;
+      }
+
+      //unconstrain parameter
+      Index idxp = nP_ + idxwmax;
+      NNLSUnconstrainParameter(idxp);
+
+    }
+
+    while (true) {
+      if (nP_==0) break;     
+
+      ampvecpermtest_ = ampVec_;
+      
+      eigen_solve_submatrix(aTaMat_,aTbVec_,ampvecpermtest_,nP_);
+
+      //check solution
+      bool positive = true;
+      for (unsigned int i = 0; i < nP_; ++i)
+        positive &= (ampvecpermtest_(i) > 0);
+      if (positive) {
+        ampVec_.head(nP_) = ampvecpermtest_.head(nP_);
+        break;
+      } 
+
+      //update parameter vector
+      Index minratioidx=0;
+      
+      // no realizable optimization here (because it autovectorizes!)
+      double minratio = std::numeric_limits<double>::max();
+      for (unsigned int ipulse=0; ipulse<nP_; ++ipulse) {
+	if (ampvecpermtest_.coeff(ipulse)<=0.) {
+	  const double c_ampvec = ampVec_.coeff(ipulse);
+	  const double ratio = c_ampvec/(c_ampvec-ampvecpermtest_.coeff(ipulse));
+	  if (ratio<minratio) {
+	    minratio = ratio;
+	    minratioidx = ipulse;
+	  }
+	}
+      }
+      ampVec_.head(nP_) += minratio*(ampvecpermtest_.head(nP_) - ampVec_.head(nP_));
+      
+      //avoid numerical problems with later ==0. check
+      ampVec_.coeffRef(minratioidx) = 0.;
+      
+      NNLSConstrainParameter(minratioidx);
+    }
+   
+    ++iter;
+
+    //adaptive convergence threshold to avoid infinite loops but still
+    //ensure best value is used
+    if (iter%10==0) {
+      threshold *= 10.;
+    }
+    
+    break;
```

Comment:
this part was marked by the static analyzer: if the "break" here is needed, the outer ```while(true)``` is not needed because there are no actual iterations here.